File: python/platform_convert.py
"""
Convert a save's raw bytes between platforms, using the project's own
bundled reference saves for the target header/size - the same approach
validated in main.py, wrapped as a pure function for reuse by the GUI.
"""

import os

from checksum import Game
from save_data import HEADER_LEN

_BASE = os.path.dirname(os.path.abspath(__file__))

# Real save files already in this repo, used as the header/size source for
# each (game, target platform) pair - see README/CHANGELOG for how these
# were validated. No confirmed real FFX2 Vita sample yet, so that one
# falls back to the PC reference (best available guess; PC and Vita are
# confirmed to share format for FFX, unconfirmed but assumed for FFX2).
REFERENCE_FILES = {
    (Game.FFX, "switch"): os.path.join(_BASE, "reference", "switch", "ffx_001"),
    (Game.FFX, "pc"): os.path.join(_BASE, "reference", "pc", "ffx_004"),
    (Game.FFX, "vita"): os.path.join(_BASE, "reference", "vita", "ffx_004"),
    (Game.FFX2, "switch"): os.path.join(_BASE, "reference", "switch", "ffx2_main_001"),
    (Game.FFX2, "pc"): os.path.join(_BASE, "reference", "pc", "ffx2_001"),
    (Game.FFX2, "vita"): os.path.join(_BASE, "reference", "pc", "ffx2_001"),
}


def _fit(data: bytes, target):
    if target is None or len(data) == target:
        return data
    if len(data) < target:
        return data + b"\x00" * (target - len(data))
    excess = data[target:]
    if any(excess):
        raise ValueError(
            f"{len(data)} bytes vs target {target}; the {len(excess)} excess "
            f"bytes are not all zero, refusing to cut real data"
        )
    return data[:target]


def convert_platform(data: bytes, game: Game, src_platform: str, dst_platform: str) -> bytes:
    """Returns `data` (currently formatted for `src_platform`) reformatted
    for `dst_platform` - header added/stripped and size matched to a real
    reference save for the target."""
    if src_platform == dst_platform:
        return bytes(data)

    body = bytes(data[HEADER_LEN[src_platform] :])
    dst_header_len = HEADER_LEN[dst_platform]

    ref_path = REFERENCE_FILES.get((game, dst_platform))
    header = b""
    target_size = None
    if ref_path and os.path.isfile(ref_path):
        with open(ref_path, "rb") as f:
            ref = f.read()
        target_size = len(ref)
        header = ref[:dst_header_len]

    return _fit(header + body, target_size)

File: typescript/src/lib/platformConvert.ts
// Convert a save's raw bytes between platforms, using this project's own
// bundled reference saves (public/reference/) for the target header/size -
// a straight port of the Python project's platform_convert.py, adapted to
// fetch its reference files instead of reading them off local disk.

import { Game } from "./checksum";
import { HEADER_LEN, type Platform } from "./saveData";

// Real save files bundled under public/reference/ - see README.md for how
// these were validated. No confirmed real FFX2 Vita sample yet, so that
// one falls back to the PC reference (best available guess; PC and Vita
// are confirmed to share format for FFX, unconfirmed but assumed for FFX2).
const REFERENCE_URLS: Record<string, string> = {
  [refKey(Game.FFX, "switch")]: "/reference/switch/ffx_001",
  [refKey(Game.FFX, "pc")]: "/reference/pc/ffx_004",
  [refKey(Game.FFX, "vita")]: "/reference/vita/ffx_004",
  [refKey(Game.FFX2, "switch")]: "/reference/switch/ffx2_main_001",
  [refKey(Game.FFX2, "pc")]: "/reference/pc/ffx2_001",
  [refKey(Game.FFX2, "vita")]: "/reference/pc/ffx2_001",
};

function refKey(game: Game, platform: Platform): string {
  return `${game}:${platform}`;
}

const referenceCache = new Map<string, Uint8Array>();

async function loadReference(game: Game, platform: Platform): Promise<Uint8Array | null> {
  const key = refKey(game, platform);
  const cached = referenceCache.get(key);
  if (cached) return cached;
  const url = REFERENCE_URLS[key];
  if (!url) return null;
  const res = await fetch(url);
  if (!res.ok) return null;
  const bytes = new Uint8Array(await res.arrayBuffer());
  referenceCache.set(key, bytes);
  return bytes;
}

function fit(data: Uint8Array, target: number | null): Uint8Array {
  if (target === null || data.length === target) return data;
  if (data.length < target) {
    const out = new Uint8Array(target);
    out.set(data);
    return out;
  }
  const excess = data.subarray(target);
  if (excess.some((b) => b !== 0)) {
    throw new Error(
      `${data.length} bytes vs target ${target}; the ${excess.length} excess bytes are not all zero, refusing to cut real data`,
    );
  }
  return data.subarray(0, target);
}

/** Returns `data` (currently formatted for `srcPlatform`) reformatted for
 * `dstPlatform` - header added/stripped and size matched to a real
 * reference save for the target. */
export async function convertPlatform(
  data: Uint8Array,
  game: Game,
  srcPlatform: Platform,
  dstPlatform: Platform,
): Promise<Uint8Array> {
  if (srcPlatform === dstPlatform) return new Uint8Array(data);

  const body = data.subarray(HEADER_LEN[srcPlatform]);
  const dstHeaderLen = HEADER_LEN[dstPlatform];

  const ref = await loadReference(game, dstPlatform);
  let header: Uint8Array = new Uint8Array(0);
  let targetSize: number | null = null;
  if (ref) {
    targetSize = ref.length;
    header = ref.subarray(0, dstHeaderLen);
  }

  const combined = new Uint8Array(header.length + body.length);
  combined.set(header, 0);
  combined.set(body, header.length);

  return fit(combined, targetSize);
}
